ivy-proofs/network_shim.ivy

```diff
@@ -0,0 +1,140 @@
+#lang ivy1.7
+# ---
+# layout: page
+# title: Network model and network shim
+# ---
+
+# Here we define a network module, which is our model of the network, and a
+# shim module that sits on top of the network and which, upon receiving a
+# message, calls the appropriate protocol handler.
+
+include domain_model
+
+# Here we define an enumeration type for identifying the 3 different types of
+# messages that nodes send.
+object msg_kind = { # TODO: merge with step_t
+    type this = {proposal, prevote, precommit}
+}
+
+# Here we define the type of messages `msg`. Its members are structs with the fields described below.
+object msg = {
+    type this = struct {
+        m_kind : msg_kind,
+        m_src  : node,
+        m_round : round,
+        m_value : value,
+        m_vround : round
+    }
+}
+
+# This is our model of the network:
+trusted isolate net = {
+# `trusted` indicates to Ivy that this is a trusted object, i.e. we assume
+# without proof that receive is only called when its requirement is true.
+
+    export action recv(dst:node,v:msg)
+    action send(src:node,dst:node,v:msg)
+    # Note that the `recv` action is exported, meaning that it can be called
+    # non-deterministically by the environment any time it is enabled. In other
+    # words, a packet that is in flight can be received at any time. In this
+    # sense, the network is fully asynchronous. Moreover, there is no
+    # requirement that a given message will be received at all.
+
+    specification {
+
+        # The state of the network consists of all the packets that have been
+        # sent so far, along with their destination.
+        relation sent(V:msg, N:node)
+
+        after init {
+            sent(V, N) := false
+        }
+
+        before send {
```

Comment:
should it be "after"?

Reply:
With `before`, the code is inserted at the beginning of the action. With `after`, it would be inserted at the end of the action. Since the action is atomic, it doesn't change much. It could make a difference if some assertions inside the action depended on the the specification state being updated or not, but I think it's not the case here.